### PR TITLE
Refactor mem_pool_host deallocation to main thread

### PR DESCRIPTION
**Pull request by pansicheng**
## Motivation

Currently, host indices pre-allocated for prefetch storage are freed by multiple threads, causing inconsistencies in mem_pool_host across different ranks. This requires synchronization operations between tp groups.

This PR proposes to centralize the freeing of host indices in the main thread, minimizing the need for cross-tp synchronization.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/srt/mem_cache/hiradix_cache.py [hunk 1]
```diff
@@ -424,6 +425,63 @@ def check_backup_progress(self):
...
+            free_request_ids = list(self.cache_controller.prefetch_free_dict.keys())
+        free_request_ids = [
+            request_id
+            for request_id in free_request_ids
+            if request_id not in self.ongoing_prefetch
+        ]
+
+        if self.tp_world_size > 1:
+            list_size = torch.tensor(len(free_request_ids), dtype=torch.int)
+            list_sizes = [torch.zeros_like(list_size) for _ in self.tp_world_size]
```
Comment (xiaguan) [inline python/sglang/srt/mem_cache/hiradix_cache.py]:
```suggestion
            list_sizes = [torch.zeros_like(list_size) for _ in range(self.tp_world_size)]
```

File: python/sglang/srt/mem_cache/hiradix_cache.py [hunk 2]
```diff
@@ -424,6 +425,63 @@ def check_backup_progress(self):
...
+            max_list_size = max(list_sizes)
+            min_list_size = min(list_sizes)
+            if min_list_size == 0:
+                free_request_ids.clear()
+            else:
+                list_value = torch.tensor(
+                    free_request_ids + [0] * (max_list_size - list_size),
+                    dtype=torch.int,
+                )
+                list_values = [torch.zeros_like(list_value) for _ in self.tp_world_size]
```
Comment (xiaguan) [inline python/sglang/srt/mem_cache/hiradix_cache.py]:
```suggestion
                list_values = [torch.zeros_like(list_value) for _ in range(self.tp_world_size)]
```

File: python/sglang/srt/mem_cache/hiradix_cache.py [hunk 3]
```diff
@@ -501,10 +556,8 @@ def check_prefetch_progress(self, req_id: str) -> bool:
         if len(written_indices):
             self.cache_controller.mem_pool_host.update_prefetch(written_indices)
 
-        self.cache_controller.mem_pool_host.free(host_indices[:matched_length])
```
Comment (xiezhq-hermann) [inline python/sglang/srt/mem_cache/hiradix_cache.py:511]: I think we can still release this here, since matched_length should be synchronized across workers.

Comment (pansicheng) [reply python/sglang/srt/mem_cache/hiradix_cache.py:511]: I’ve made the changes, please take a look!

**Comment by pansicheng** [comment]
to replace this with https://github.com/sgl-project/sglang/pull/9797
